crates/dynamodb-streams/src/stream.rs

```diff
@@ -0,0 +1,169 @@
+            match result {
+                Ok((next_iter, records)) => {
+                    if let Some(batch) =
+                        self.state.handle_poll_result(&shard_id, next_iter, records)
+                    {
+                        batches.push(batch);
+                    }
+                }
+                Err(e) => {
+                    tracing::error!("Shard poll failed: shard_id={}, {}", shard_id, e);
+                    self.handle_failed_shard(&shard_id, &e);
+                }
+            }
+        }
+
+        // 3. Discover new shards
+        if let Ok(shards) = self.client.get_all_shards(&self.stream_arn).await {
+            self.state.add_discovered(shards);
+        }
+
+        // 4. Initialize shards that require iterators
+        self.initialize_shards_iterators().await;
+
+        Ok(batches)
+    }
+
+    fn handle_failed_shard(&mut self, shard_id: &str, error: &Error) {
+        // TODO: Finalize logic
+        let is_expired_iterator = error.to_string().contains("ExpiredIterator")
+            || error.to_string().contains("TrimmedDataAccess");
+
+        if is_expired_iterator {
+            tracing::warn!("Iterator expired for shard {}, will reinitialize", shard_id);
+
+            self.state.reinitialize_shard(shard_id);
+        }
+    }
+
+    async fn initialize_shards_iterators(&mut self) {
+        let shard_ids: Vec<String> = self.state.get_initializing_shards_ids().cloned().collect();
+
+        for shard_id in shard_ids {
+            match self
+                .client
+                .get_shard_iterator(
+                    &self.stream_arn,
+                    &shard_id,
+                    &ShardIteratorType::TrimHorizon,
+                    None,
+                )
+                .await
+            {
+                Ok(iterator) => {
+                    if let Some(iterator) = iterator {
+                        self.state.mark_active(shard_id, iterator);
+                    }
+                }
+                Err(e) => {
+                    tracing::warn!("Failed to initialize shard {}: {}", shard_id, e);
+                }
+            }
+        }
+    }
+
+    async fn perform_iterate_with_retry(&mut self) -> Result<Vec<RecordBatch>> {
+        let mut backoff = self.retry_strategy.clone();
+
+        loop {
+            match self.iterate().await {
+                Ok(result) => return Ok(result),
+                Err(e) => {
+                    if let Some(duration) = backoff.next_backoff() {
+                        tracing::debug!("Iteration failed, retrying after {:?}: {}", duration, e);
+                        tokio::time::sleep(duration).await;
+                    } else {
+                        tracing::error!("Iteration failed after exhausting retries: {}", e);
+                        return Err(e);
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn streaming(mut self) {
```

Comment:
This isn't really streaming? It still collects the entire shard into memory before sending the batches, for a true streaming we need a way to yield batches.